README
       mfit: Model Fitting Software for Optical Interferometry

                       package release 1.5.2


                           Release notes
			   
LICENSING

This software is released into the public domain, in the hope that it
will be useful. There is of course no warranty of any sort.


DESCRIPTION

mfit is a tool for fitting simple parametrisations of the sky
brightness distribution to optical (visible/IR) interferometry data.

Features of mfit include:

* Asymmetric, multi-component models 
* Four limb-darkening parametrizations (Taylor and Gauss-Hermite
  expansions, Hestroffer and square-root models), plus support for arbitrary
  limb-darkening profiles specified numerically (may be wavelength-dependent)
* Support for wavelength-dependent models  (specified model parameters
  can take different values for different wavebands)
* Interactive, command-line, and GUI interfaces 
* Plots of data and model
* Optional marginalisation by numerical integration of the
  posterior probability (requires commercial NAG library)

The graphical interface, fitgui, can straightforwardly be integrated
into other applications written in python.

Please refer to the NEWS file for details of user-visible changes in this
release.


REQUIREMENTS

mfit should work on any unix-like operating system (see below for those it has
been tested on). You will need a Fortran 90 or 95 compiler to build mfit.

mfit must be linked with the following libraries (not supplied, but free
software):

* PGPLOT - http://www.astro.caltech.edu/~tjp/pgplot/

* CFITSIO - http://heasarc.gsfc.nasa.gov/docs/software/fitsio/fitsio.html

* FFTW (v3) - http://www.fftw.org/

If you have it, mfit can make use of the commercial Numerical Algorithms Group
Fortran Library (Mark 19 recommended) - see
http://www.nag.co.uk/numeric/fl/FLdescription.asp

If the NAG library is not available, the marginalisation-by-integration
features of mfit ('--margerr', '--plot mpost', '--plot mpost2d' in clfit) will
not work.

To build mfit, you will need scons - see http://www.scons.org - and a working
installation of Python (1.5.2 or later).

To run the fitgui graphical interface, you will need a working
installation of Python 2.2 or above and Tkinter (see
http://www.python.org).


FILES

You can download the files from ftp://ftp.mrao.cam.ac.uk - login as
anonymous and get everything in pub/jsy1001/mfit.

Alternatively follow the link from
http://www.mrao.cam.ac.uk/~jsy1001/mfit/

The files in this package are as follows:

ChangeLog       Package ChangeLog (automatically generated from CVS
		log messages)
README	        This file
NEWS            Outline of new features in this and previous releases
documentation   Instructions for using mfit
SConstruct      scons configuration file, used to build mfit (see below)
bayes.f90, fit.f90, inout.f90, marginalise.F90, model.f90, plot.f90, postplot.f90,
visibility.f90, wrap.f90
		Routines used by mfit and clfit
main.f90	mfit main function
clfit.f90	clfit main function
calc.f90	Test code for maths module
modelplot.f90, fitsimage.f90
	        Image-space model plot utility
fitgui.py	GUI interface to clfit
fitgui		Wrapper script for fitgui.py - put this in PATH
fitgui_dev      Alt. wrapper script for fitgui.py - used for development
test.oifits     Test dataset (COAST data on Capella)
test.model      Binary model that fits test.oifits
pda_xermsg.f    Replacement for Starlink routine

The following files, authored by third parties, are distributed with mfit for
your convenience:

gamma.f	        Gamma function from SPECFUN 2.5 - see readme.specfun
rjbesl.f        Bessel function from SPECFUN 2.5 - see readme.specfun
readme.specfun	Readme for SPECFUN 2.5 routines
gmst.f, dranrm.f
                Sidereal time function from SLALIB
maths_pda.f, fit_pda.f
                Functions from Starlink PDA library
f2kcli.f90	Fortran 200x Command Line Interface, (c) Interactive
		Software Services Ltd. - see f2kcli.txt
f2kcli_nagw.f90	Fortran 200x Command Line Interface - version for NAGWare f95
f2kcli.txt	manual.txt from f2kcli distribution


BUILDING MFIT

scons is now used to build mfit. If the 'scons' command is not available on
your system, retrieve scons from http://www.scons.org and install it.

If your combination of operating system and Fortran 90/95 compiler is not one
of those listed below, you *may* need to edit the section labelled
'#### May need to edit this section ####' at the top of the SConstruct file.

Note that the SConstruct file is a Python script. Please contact the author if
the distributed SConstruct file does not work for you and you are not sure how
to edit it.

Build mfit by typing 'scons' at the shell prompt (from the directory containing
the source code). The following command-line options are supported:
'debug=1': compile with debugging flags (-g etc.) - this is the default
'debug=0': compile without debugging flags
'release=0': compile without optimization flags - this is the default
'release=1': compile with  optimization flags (-O etc.)
'f95=/path/to/f95': use specified compiler, rather than auto-detected one


mfit has been tested using:
- Sun Workshop Fortran 95 on Solaris
- NAGWare Fortran 95 on Solaris
- NAGWare Fortran 95 on Linux
- G95 on Linux
- GFortran 4.3 on Linux (will not work with older GFortran versions)
- G95 on MacOS X (thanks to Ettore Pedretti)

Please report successes or failures with other compilers/operating systems.


INSTALLING MFIT

Build the executables as described above.

Edit the command following '#!' at the top of 'fitgui' so that it
starts the Python interpreter on your system. Put the 'mfit' and
'clfit' executables and the 'fitgui' script somewhere in the PATH. Put
'fitgui.py' somewhere in your python module search path (e.g. in one
of the directories in $PYTHONPATH).

[If you are a COASTie, the following commands should accomplish the
 above for our complicated install scheme (make sure you have checked
 the 'build' package out of CVS; refer to the Data Reduction Manual,
 'Maintaining the Software' and
 http://oberon.ra.phy.cam.ac.uk/coast/local_web/rsync.html):

 > gmake -f makefile.coast install
 > epkg -s DEPOT-DIR/mfit mfit

 where DEPOT-DIR is /coast/depot-sparc if installing on the Sun cluster, or
 /coast/depot-i386 for Linux.
]


USING MFIT

Please refer to the instructions in the 'documentation' file.


BUGS

Please report the inevitable bugs to jsy1001 (at) cam.ac.uk. Various
missing features are listed in 'documentation'.


$Id: README,v 1.5 2009/11/03 17:26:26 jsy1001 Exp $

Local variables:
fill-column: 79
End:


NEWS

New in mfit release 1.5.3
=========================

* Plotted uv points are saved to text files (this feature was
  previously added for other plot types)


New in mfit release 1.5.2
=========================

* No longer depends on Starlink SLALIB and PDA libraries; the required
  source code from these libraries is distributed with mfit.


New in mfit release 1.5.1
=========================

* Minimiser now works with scaled model parameters, hence performs
  better. Parameter scaling factors are varied iteratively to find the
  best minimum

* Now built using scons (more portable) rather than make

* Use of NAG library optional (detected by scons)

* Support for brute-force marginalisation - integrates posterior
  probability in multiple dimensions (slow, requires NAG library)

* Support for wavelength-dependent models (specified model parameters
  can take different values for different wavebands)

* Support for rotating model components (i.e. polar angle theta varies
  linearly with time)

* '#relto' option for model component positions - specifies position
  of model component w.r.t. position of another component

* Comments now allowed anywhere in model file

* Added plots versus Modified Julian Day, GMST and baseline position angle

* Plotted points are also saved to text files, for re-plotting in
  other applications

* Added 'mplot' application to plot intensity distribution of model


New in mfit release 1.4.1
=========================

* Replaced Numerical Recipes routines with free equivalents; mfit can
  now be distributed to the community

* Added support for *wavelength-dependent* numerical centre-to-limb
  intensity profiles

* User can specify a range of wavelengths to use from the data

* Added plots of squared visibility and triple amplitude and phase
  versus wavelength

* User can specify a TARGET_ID to use from OIFITS data

* User can specify calibration error (extra fractional error in mod V)
  to add to OIFITS data. This feature was previously restricted to
  other formats


New in mfit release 1.3.1 (released to COAST team on 2003/09/11)
=========================

* Added support for numerical centre-to-limb intensity profiles e.g.
Kurucz models

* Takes into account bandwidth smearing (assuming a tophat bandpass)

* Option to plot 1-d cut through posterior

This version *may not* be released to the public, as
non-freely-distributable code from Numerical Recipes is used (other
routines could be substituted if this were deemed to be worth the effort).
The public website for mfit - http://www.mrao.cam.ac.uk/~jsy1001/mfit/ -
describes the previous 1.2.1 release.


mfit release 1.2.1 (public release on 2003/07/18)
==================

Previous public release.


fitgui.py
#!/usr/bin/env python
#
# $Id: fitgui.py,v 1.13 2008/05/14 15:11:51 jsy1001 Exp $

"""Graphical user interface for clfit.

Usage: fitgui &

"""


import string, sys, os, tempfile, popen2, fcntl, time
from Tkinter import *
from ScrolledText import ScrolledText
import tkFileDialog

_revision = string.split("$Revision: 1.13 $")[1]


class GUI:
    """Graphical user interface to clfit.

    Data attributes that may sensibly be used externally:

    exe (string) -- path to clfit executable

    device (string) -- PGPLOT device, passed to clfit

    fileName (Tkinter.StringVar) -- path to OI-FITS/Mapping Data file
    (use set() & get() methods)

    ChangeFileButton (Tkinter.Button) -- brings up data file dialog
    box (may wish to disable)
    
    initialdir (string) -- initial directory for file dialog boxes

    preFitCallback -- function to call just before running clfit. If this
    returns false, clfit won't be run

    postFitCallback -- function to call after running clfit (if successful)
    
    """

    def __init__(self, parent, dismissCommand=None):
        """Constructor.

        Arguments are:

        parent -- parent window

        dismissCommand -- function to call when Dismiss button clicked

        """
        # Initialise data attributes
        self.parent = parent
        self.exe = 'clfit'
        self.device = '/xserv'
        self.fileName = StringVar()
        self.fileName.set('(unset)')
        self.initialdir = os.getcwd()
        self.preFitCallback = None
        self.postFitCallback = None
        self.calErr = StringVar()
        self.calErr.set('0.0')
        self.cwl = StringVar()
        self.bw = StringVar()
        self.wlmin = StringVar()
        self.wlmax = StringVar()
        self.target_id = StringVar()
        self.nofit = IntVar()
        self.nofit.set(0)
        self.plots = ['No plot', 'uv', 'vis2', 't3amp', 't3phi',
                      'vis2-wl', 't3amp-wl', 't3phi-wl',
                      'vis2-mjd', 't3amp-mjd', 't3phi-mjd',
                      'vis2-st', 't3amp-st', 't3phi-st', 'vis2-pa',
                      'post', 'mpost', 'post2d', 'mpost2d']
        self.selPlot = StringVar()
        self.selPlot.set(self.plots[1])
        self.plotXIndex = StringVar()
        self.plotXIndex.set('1')
        self.plotYIndex = StringVar()
        self.plotYIndex.set('2')
        self.plotXFrom = StringVar()
        self.plotXTo = StringVar()
        self.plotYFrom = StringVar()
        self.plotYTo = StringVar()
        self.margErr = IntVar()
        self.margErr.set(0)
        self.margErrVar = StringVar()
        self.margErrVar.set('1')

        # Initialise GUI elements
        fileFrame = Frame(parent)
        fileFrame.pack(side=TOP)
        Label(fileFrame, text='Data file:').pack(side=LEFT)
        Label(fileFrame, textvariable=self.fileName).pack(side=LEFT)
        self.ChangeFileButton = Button(fileFrame, text='Change',
                                       command=self._ChangeFileName)
        self.ChangeFileButton.pack(side=LEFT)
        calErrFrame = Frame(parent)
        calErrFrame.pack(side=TOP, fill=X, pady=4)
        Label(calErrFrame, text='Calibration Error (extra frac. error in system vis.)').pack(side=LEFT, anchor=W)
        Entry(calErrFrame, textvariable=self.calErr, width=5).pack(
            side=LEFT, anchor=W, padx=4)
        wbFrame = Frame(parent)
        wbFrame.pack(side=TOP, fill=X, pady=4)
        Label(wbFrame, text='Waveband:').pack(side=LEFT, anchor=W)
        Entry(wbFrame, textvariable=self.cwl, width=5).pack(side=LEFT,
                                                            anchor=W, padx=4)
        Entry(wbFrame, textvariable=self.bw, width=5).pack(side=LEFT,
                                                           anchor=W, padx=4)
        Label(wbFrame, text='or Wavelength range:').pack(side=LEFT, anchor=W)
        Entry(wbFrame, textvariable=self.wlmin,
              width=5).pack(side=LEFT, anchor=W, padx=4)
        Entry(wbFrame, textvariable=self.wlmax,
              width=5).pack(side=LEFT, anchor=W, padx=4)
        targetFrame = Frame(parent)
        targetFrame.pack(side=TOP, fill=X, pady=4)
        Label(targetFrame, text='TARGET_ID (blank to use 1st in OI_TARGET table):').pack(side=LEFT, anchor=W)
        Entry(targetFrame, textvariable=self.target_id,
              width=5).pack(side=LEFT, anchor=W, padx=4)
        Label(parent, text='Model:').pack(side=TOP, anchor=W)
        self.ModelText = ScrolledText(parent, height=19, width=40,
                                      font=('Helvetica', 10))
        self.ModelText.pack(side=TOP, expand=1, fill=BOTH)
        midFrame1 = Frame(parent)
        midFrame1.pack(side=TOP, fill=X, pady=4)
        Label(midFrame1, text='Plot:').pack(side=LEFT, anchor=NW)
        plotFrame = Frame(midFrame1)
        plotFrame.pack(side=LEFT)
        for i in range(len(self.plots)):
            p = self.plots[i]
            Radiobutton(plotFrame, text=p, variable=self.selPlot,
                        value=p).grid(row=(i+1)/2, column=(i+1)%2, sticky=W)
        Entry(plotFrame, textvariable=self.plotXIndex,
              width=3).grid(row=len(self.plots)/2-1, column=2)
        Entry(plotFrame, textvariable=self.plotYIndex,
              width=3).grid(row=len(self.plots)/2, column=2)
        rangeFrame = Frame(midFrame1)
        rangeFrame.pack(side=LEFT)
        Label(rangeFrame, text='X From:').grid(row=0, column=0, sticky=E)
        Entry(rangeFrame, textvariable=self.plotXFrom, width=5).grid(row=0, column=1)
        Label(rangeFrame, text='To:').grid(row=0, column=2)
        Entry(rangeFrame, textvariable=self.plotXTo, width=5).grid(row=0, column=3)
        Label(rangeFrame, text='[(m)post2d only] Y From:').grid(row=1, column=0, sticky=E)
        Entry(rangeFrame, textvariable=self.plotYFrom, width=5).grid(row=1, column=1)
        Label(rangeFrame, text='To:').grid(row=1, column=2)
        Entry(rangeFrame, textvariable=self.plotYTo, width=5).grid(row=1, column=3)
        Button(midFrame1, text='Go', command=self.Go).pack(side=RIGHT,
                                                           anchor=NE, padx=4)
        Button(midFrame1, text='Save model',
               command=self.SaveModel).pack(side=RIGHT, anchor=NE, padx=4)
        Button(midFrame1, text='Load model',
               command=self.LoadModel).pack(side=RIGHT, anchor=NE, padx=4)
        midFrame2 = Frame(parent)
        midFrame2.pack(side=TOP, fill=X, pady=4)
        Checkbutton(midFrame2, text="Don't fit (report goodness-of-fit only)",
                    variable=self.nofit).pack(side=LEFT, anchor=W, padx=8)
        Entry(midFrame2, textvariable=self.margErrVar,
              width=5).pack(side=LEFT, anchor=W)
        Checkbutton(midFrame2, text="Error bar by marginalising",
                    variable=self.margErr).pack(side=LEFT, anchor=W)
        midFrame3 = Frame(parent)
        midFrame3.pack(side=TOP, fill=X)
        Label(midFrame3, text='Results:').pack(side=LEFT, anchor=SW)
        if dismissCommand is None: dismissCommand = parent.quit
        Button(midFrame3, text='Dismiss',
               command=dismissCommand).pack(side=RIGHT, padx=4, pady=4)
        Button(midFrame3, text='Clear results',
               command=self.ClearResults).pack(side=RIGHT, padx=4, pady=4)
        self.Results = ScrolledText(parent, height=31, width=90,
                                    font=('Courier', 10), state=DISABLED)
        self.Results.tag_config('result', foreground='#1e90ff') # dodger blue
        self.Results.tag_config('commentary', foreground='#ff8c00') # dark orange
        self.Results.tag_config('error', foreground='#8b0000') # dark red
        self.Results.pack(side=TOP, expand=1, fill=BOTH)

    def LoadModel(self):
        """Get filename and read model from file."""
        fileName = tkFileDialog.askopenfilename(parent=self.parent,
            initialdir=self.initialdir,
            filetypes=[('mfit model files','*.model'),
                       ('All files','*')])
        if fileName != '': self.ReadModel(fileName)

    def ReadModel(self, fileName):
        """Read model from file."""
        try:
            fil = file(fileName, 'r')
            text = fil.read()
            fil.close()
        except IOError, (errNo, errStr):
            self.ShowResult('Error reading %s: %s\n' % (fileName, errStr),
                            'error')
        else:
            self.SetModel(text)

    def SetModel(self, text):
        """Set model text."""
        self.ModelText.delete(1.0, END)
        self.ModelText.insert(END, text)

    def ClearResults(self):
        """Clear results window."""
        self.Results.configure(state=NORMAL)
        self.Results.delete(1.0, END)
        self.Results.configure(state=DISABLED)

    def SaveModel(self):
        """Get filename and write model to file."""
        fileName = tkFileDialog.asksaveasfilename(parent=self.parent,
            initialdir=self.initialdir,
            filetypes=[('mfit model files','*.model'),
                       ('All files','*')])
        if fileName != '': self.WriteModel(fileName)

    def WriteModel(self, fileName):
        """Write model text to file."""
        fil = file(fileName, 'w')
        fil.write(self.ModelText.get(1.0, END))
        fil.close()

    def _ChangeFileName(self):
        newName = tkFileDialog.askopenfilename(
            parent=self.parent,
            initialdir=self.initialdir,
            title='Choose data file for fit',
            filetypes=[('(OI-)FITS files','*fits'), ('COAST Mapping Data files', '*.mapdat'),
                       ('wbCalib / nbCalib files','*calib'), ('All files','*')])
        if newName != '': self.fileName.set(newName)

    def ShowResult(self, text, tag):
        """Display text in 'Results'."""
        self.Results.configure(state=NORMAL)
        self.Results.insert(END, text, tag)
        self.Results.yview(END)
        self.Results.configure(state=DISABLED)
        self.parent.update_idletasks()

    def Go(self):
        """Fit the current model."""
        # Execute pre-callback
        if callable(self.preFitCallback):
            if not self.preFitCallback(): return
        # Write model text to tempfile
        self._tempName = tempfile.mktemp('.model')
        self.WriteModel(self._tempName)
        # Run clfit so we can grab its output
        optText = ''
        try:
            c = float(self.calErr.get())
        except ValueError:
            pass
        else:
            optText += ' --calerr %.3f' % c
        try:
            cwl = float(self.cwl.get())
            bw = float(self.bw.get())
        except ValueError:
            try:
                wlmin = float(self.wlmin.get())
                wlmax = float(self.wlmax.get())
                optText += ' --waverange %.2f %.2f' % (wlmin, wlmax)
            except ValueError:
                pass # don't specify waveband(s)
        else:
            optText += ' --waveband %.2f %.2f' % (cwl, bw)
        try:
            target_id = int(self.target_id.get())
            optText += ' --target_id %d' % target_id
        except ValueError:
            pass
        p = self.selPlot.get()
        if p != self.plots[0]: # not 'No plot'
            if p == 'post' or p == 'mpost':
                try:
                    index = int(self.plotXIndex.get())
                except ValueError:
                    index = 1
            if p == 'post2d' or p == 'mpost2d':
                try:
                    indx = (int(self.plotXIndex.get()),
                            int(self.plotYIndex.get()))
                except ValueError:
                    indx = (1, 2)
            try:
                xmin = float(self.plotXFrom.get())
                xmax = float(self.plotXTo.get())
                if p[-2:] == '2d':
                    ymin = float(self.plotYFrom.get())
                    ymax = float(self.plotYTo.get())
            except ValueError:
                if p == 'post' or p == 'mpost':
                    optText += ' --plot %s %d' % (p, index)
                elif p == 'post2d' or p == 'mpost2d':
                    optText += ' --plot %s %d %d' % (p, indx[0], indx[1])
                else:
                    optText += ' --plot %s' % p
            else:
                if p == 'post' or p == 'mpost':
                    optText += ' --zoomplot %s %d %.4f %.4f' % (p, index, xmin, xmax)
                elif p == 'post2d' or p == 'mpost2d':
                    optText += ' --zoomplot %s %d %d %.4f %.4f  %.4f %.4f' % \
                               (p, indx[0], indx[1], xmin, xmax, ymin, ymax)
                else:
                    optText += ' --zoomplot %s %.4f %.4f' % (p, xmin, xmax)
        if self.nofit.get(): optText += ' --nofit'
        if self.margErr.get():
            optText += ' --margerr %s' % (self.margErrVar.get())
        command = 'nice %s%s --device %s %s %s' % (
            self.exe, optText, self.device, self.fileName.get(),
            self._tempName)
        self.ShowResult('Running %s:\n' % command, tag='commentary')
        self._popen4 = popen2.Popen4(command, 10)
        fd = self._popen4.fromchild.fileno() # stdout & stderr of child process
        oldflags = fcntl.fcntl(fd, fcntl.F_GETFL)
        fcntl.fcntl(fd, fcntl.F_SETFL, oldflags|os.O_NONBLOCK)
        self.parent.tk.createfilehandler(fd, READABLE, self._HandleChildOutput)

    def _HandleChildOutput(self, fd, mask):
        """Handle output from child process."""
        self._ShowOutput()
        status = self._popen4.poll()
        if status != -1:
            # child process completed
            time.sleep(0.5) # wait for last output
            self._ShowOutput()
            if os.WEXITSTATUS(status) != 0:
                self.ShowResult('Subprocess exited with code %d\n' \
                                % os.WEXITSTATUS(status), 'error')
            else:
                self.ShowResult('Subprocess exited normally\n', 'commentary')
            self.parent.tk.deletefilehandler(fd)
            os.remove(self._tempName)
            # Execute post-callback
            if callable(self.postFitCallback):
                self.postFitCallback()

    def _ShowOutput(self):
        """Read child process output and pass to ShowResult()."""
        try:
            result = self._popen4.fromchild.read()
        except IOError, (errNo, errMsg):
            self.ShowResult('I/O Error %d: %s\n' % (errNo, errMsg),
                            'commentary')
        else:
            self.ShowResult(result, 'result')
    

def _main(altExe=None):
    """Main routine."""

    # Find out if Python Megawidgets are installed
    try:
        import Pmw
    except ImportError:
        havePmw = 0
    else:
        havePmw = 1

    if len(sys.argv) == 1:
        # No command-line arguments - run graphical user interface
        root = Tk()
        if havePmw:
            Pmw.initialise(root, fontScheme='pmw1')
        root.title('fitgui %s' % _revision)
        main = GUI(root)
        main.fileName.set('test.oifits')
        main.ReadModel('test.model')
        if altExe is not None: main.exe = altExe
        root.mainloop()
    else:
        # Too many arguments
        print "%s %s\n" % (sys.argv[0], _revision)
        print __doc__
        sys.exit(2)

if __name__ == '__main__':
    _main()
 
# Local Variables:
# mode: python
# End:
